pkg/kv/kvserver/store_snapshot.go

```diff
@@ -427,54 +427,61 @@ func (kvSS *kvBatchSnapshotStrategy) Send(
 		return nil
 	}
 
-	iter := snap.Iter
-	var ok bool
-	var err error
-	for ok, err = iter.SeekStart(); ok && err == nil; ok, err = iter.Next() {
-		// NB: EngineReplicaDataIterator will never expose point/range keys
-		// together: it first iterates over all point keys in a key range, then over
-		// all range keys in a key range, then moves onto the next key range.
-		hasPoint, hasRange := iter.HasPointAndRange()
-
-		if hasPoint {
-			kvs++
-			var unsafeKey storage.EngineKey
-			if unsafeKey, err = iter.UnsafeKey(); err != nil {
-				return 0, err
-			}
-			unsafeValue := iter.UnsafeValue()
-			if b == nil {
-				b = kvSS.newBatch()
-			}
-			if err := b.PutEngineKey(unsafeKey, unsafeValue); err != nil {
-				return 0, err
-			}
-			if int64(b.Len()) >= kvSS.batchSize {
-				if err = flushBatch(); err != nil {
-					return 0, err
-				}
-			}
+	maybeFlushBatch := func() error {
+		if int64(b.Len()) >= kvSS.batchSize {
+			return flushBatch()
 		}
+		return nil
+	}
 
-		if hasRange {
-			bounds, err := iter.RangeBounds()
-			if err != nil {
-				return 0, err
-			}
-			for _, rkv := range iter.RangeKeys() {
-				rangeKVs++
-				err := b.PutEngineRangeKey(bounds.Key, bounds.EndKey, rkv.Version, rkv.Value)
-				if err != nil {
-					return 0, err
+	err := rditer.IterateReplicaKeySpans(snap.State.Desc, snap.EngineSnap, true, /* replicatedOnly */
+		func(iter storage.EngineIterator, _ roachpb.Span, keyType storage.IterKeyType) error {
+			var err error
+			switch keyType {
+			case storage.IterKeyTypePointsOnly:
+				for ok := true; ok && err == nil; ok, err = iter.NextEngineKey() {
+					kvs++
+					if b == nil {
+						b = kvSS.newBatch()
+					}
+					var key storage.EngineKey
+					if key, err = iter.UnsafeEngineKey(); err != nil {
+						return err
+					}
+					if err = b.PutEngineKey(key, iter.UnsafeValue()); err != nil {
+						return err
+					}
+					if err = maybeFlushBatch(); err != nil {
+						return err
+					}
 				}
-				if bLen := int64(b.Len()); bLen >= kvSS.batchSize {
-					if err = flushBatch(); err != nil {
-						return 0, err
+
+			case storage.IterKeyTypeRangesOnly:
+				for ok := true; ok && err == nil; ok, err = iter.NextEngineKey() {
+					bounds, err := iter.EngineRangeBounds()
+					if err != nil {
+						return err
+					}
```

Comment:
nit: maybe make it consistent with above
```
var bounds ...
if bounds, err = iter.EngineRangeBounds(); err != nil {
    return err
}
```

We also have very similar piece of code in replica_leartner_test.go

Reply:
Ended up using `:=`, since we're not worried about shadowing `err` here.